File: core/src/main/java/com/illuzionzstudios/tab/components/loader/TabLoader.java
package com.illuzionzstudios.tab.components.loader;

import com.illuzionzstudios.mist.config.ConfigSection;
import com.illuzionzstudios.tab.components.text.DynamicText;
import com.illuzionzstudios.tab.components.text.FrameText;
import com.illuzionzstudios.tab.components.text.ScrollableText;
import com.illuzionzstudios.tab.controller.TabController;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

/**
 * Created instance of a tab loader.
 * What this class does is when instantiated, it
 * loads all information from a configuration section
 * then stores it.
 */
public class TabLoader implements Loader {

    /**
     * Permission needed to view tab
     */
    @Getter
    private String permission;

    /**
     * Weight of the tab
     */
    @Getter
    private int weight;

    /**
     * The columns this tab displays.
     * These are loaded last so we can store the objects
     */
    @Getter
    private List<Loader> columns = new ArrayList<>();

    /**
     * List of header elements
     */
    @Getter
    private List<DynamicText> header = new ArrayList<>();

    /**
     * List of footer elements
     */
    @Getter
    private List<DynamicText> footer = new ArrayList<>();

    public TabLoader(ConfigSection section) {
        permission = section.getString("Permission");
        weight = section.getInt("Weight");

        // Loop through and search for loaded column loaders
        for (String column : section.getStringList("Columns")) {
            Loader loadedColumn = TabController.INSTANCE.getLoaders().get(column.toLowerCase());
            if (loadedColumn != null) {
                this.columns.add(loadedColumn);
            }
        }

        // Load header elements
        for (ConfigSection text : section.getSections("Header.Text")) {
            // Add to elements
            List<String> frames = text.getStringList("Animations");
            DynamicText element = text.getBoolean("Scroll.Enabled") ?
                    new ScrollableText(text.getInt("Scroll.Interval"), frames.get(0)) :
                    new FrameText(text.getInt("Interval"), frames);
            header.add(element);
        }

        // Load footer elements
        for (ConfigSection text : section.getSections("Footer.Text")) {
            // Add to elements
            List<String> frames = text.getStringList("Animations");
            DynamicText element = text.getBoolean("Scroll.Enabled") ?
                    new ScrollableText(text.getInt("Scroll.Interval"), frames.get(0)) :
                    new FrameText(text.getInt("Interval"), frames);
            footer.add(element);
        }
    }

}

File: core/src/main/kotlin/com/illuzionzstudios/tab/tab/components/loader/TabLoader.kt
package com.illuzionzstudios.tab.tab.components.loader

import com.illuzionzstudios.mist.Logger
import com.illuzionzstudios.mist.config.YamlConfig
import com.illuzionzstudios.mist.config.serialization.loader.YamlFileLoader
import com.illuzionzstudios.tab.model.DynamicText
import com.illuzionzstudios.tab.tab.TabController
import com.illuzionzstudios.tab.tab.components.Tab
import com.illuzionzstudios.tab.tab.components.column.TabColumn

/**
 * This is the object that loads the tab from our config. If we want
 * to edit the tab instance we edit this and re-save and load our tab object
 * to then update in game
 */
class TabLoader(directory: String, fileName: String): YamlFileLoader<Tab>(directory, fileName) {

    override fun loadYamlObject(file: YamlConfig?): Tab {
        val tab = Tab(file?.getString("name") ?: "default")

        tab.permission = file?.getString("permission") ?: ""
        tab.weight = file?.getInt("weight") ?: 1

        tab.displayTitles = file?.getBoolean("columns.display-titles", true) ?: true
        tab.elementWidth = file?.getInt("columns.width") ?: 50

        // Load columns into tab
        // COLUMNS MUST BE LOADED FIRST
        file?.getConfigurationSection("columns.list")?.getKeys(false)?.forEach {
            // Try columns then list
            tab.columns[it.toInt()] = TabController.columns[file.getString("columns.list.$it")] ?: TabController.lists[file.getString("columns.list.$it")]!!
        }

        val headerText: MutableList<DynamicText> = ArrayList()
        file?.getSections("header.text")?.forEach {
            headerText.add(DynamicTextLoader(it!!).`object`)
        }
        tab.header = headerText

        val footerText: MutableList<DynamicText> = ArrayList()
        file?.getSections("footer.text")?.forEach {
            footerText.add(DynamicTextLoader(it!!).`object`)
        }
        tab.footer = footerText

        return tab
    }

    override fun saveYaml() {
        TODO()
    }
}